maap/Result.py

```diff
@@ -23,71 +22,65 @@ def getLocalPath(self, destpath=".", overwrite=False):
+                if not overwrite and not os.path.isfile(destpath + "/" + filename):
+                    s3 = boto3.client('s3')
+                    s3.download_file(o.netloc, o.path.lstrip('/'), destpath + "/" + filename)
+            except:
+                # Fall back to HTTP
+                http_url = self._convertS3toHttp(url)
+                return self._getHttpData(http_url, overwrite, destpath, destfile)
 
             return destpath + '/' + filename
         else:
-            if not overwrite and not os.path.isfile(destpath + "/" + destfile):
-                r = requests.get(url, stream=True)
-                r.raw.decode_content = True
+            return self._getHttpData(url, overwrite, destpath, destfile)
 
-                with open(destpath + "/" + destfile, 'wb') as f:
-                    shutil.copyfileobj(r.raw, f)
-
-            return destpath + '/' + destfile
-
-    def getRelatedData(self, destpath=".", overwrite=False):
+    def getLocalPath(self, destpath=".", overwrite=False):
         """
-        Download the dataset into file system
-        :param destpath: use the current directory as default
-        :param overwrite: don't download by default if the target file exists
-        :return:
+        Deprecated method. User getData() instead.
         """
+        return self.getData(destpath, overwrite)
 
-        downloads = self._relatedUrls
-        output = []
-
-        # Downloadable url does not exist
-        if not downloads:
-            return None
-
-        for download in downloads:
-            url = download['URL']
-            destfile = url.split("/")[-1].replace('/', '')
-
-            if not overwrite and not os.path.isfile(destpath + "/" + destfile):
+    def _convertS3toHttp(self, url):
+        url = url[5:].split('/')
+        url[0] += '.s3.amazonaws.com'
+        url = 'https://' + '/'.join(url)
+        return url
 
-                # Get granule oauth2 URL
-                s = requests.Session()
-                response = s.get(url)
-                #print(response.status_code, response.url)
-                # Get granule
-                s.headers.update({'Authorization': f'Bearer {self._ursToken},Basic {os.environ.get("MAAP_APP_CREDS")}',
-                                  'Connection':'close'})
+    def _getHttpData(self, url, overwrite, destpath, destfile):
+        if not overwrite and not os.path.isfile(destpath + "/" + destfile):
+            r = requests.get(url, stream=True)
 
-                r = s.get(url=response.url, stream=True)
-                r.raw.decode_content = True
+            # Try with a federated token if unauthorized
+            if r.status_code == 401:
+                r = requests.get(
+                    url=os.path.join(self._cmrFileUrl,
+                                     urllib.parse.quote(urllib.parse.quote(url, safe='')),
+                                     endpoints.CMR_ALGORITHM_DATA),
+                    headers=self._apiHeader,
```

Comment:
I still don't understand the full federated token workflow but it seems like this method and the `_PROXY_GRANTING_TICKET` is the crux of it https://github.com/MAAP-Project/maap-py/blob/master/maap/maap.py#L83-L90 is that right?